arangod/StorageEngine/TransactionState.h

```diff
@@ -392,7 +397,7 @@ class TransactionState : public std::enable_shared_from_this<TransactionState> {
 
   // in case of read-only transactions collections can be added lazily.
   // this can happen concurrently, so for this we need to protect the list
-  mutable std::mutex _collectionsLock;
+  mutable std::shared_mutex _collectionsLock;
```

Comment:
Not sure about this, but I would have gone for a `basics::ReadWriteLock`
We also have RECURSIVE_READ_LOCKER and RECURSIVE_WRITE_LOCKER helpers
In case those methods stack and could acquire locks recursively.
As an alternative: We can make recursively called functions in a private/protected "NoLock" variant, that simply "assumes" a lock is held, all public methods however needs to setup proper locking.
A